src/v/io/cache.h

```diff
@@ -0,0 +1,662 @@
+ *
+ * The cache hook also provides an interface between the entry and the cache for
+ * communicating entry state changes such as the entry being evicted, or the
+ * application indicating that the entry has been used in a meaningful way.
+ *
+ * When the application using the cache rehydrates an entry on the ghost queue
+ * it is important to maintain the metadata contained in the cache hook. If the
+ * entry structure is re-used then it should be sufficient to avoid resetting
+ * the cache hook member. However, if the entry is destroyed and a replacement
+ * is created, then the application should initialize the hook in the new entry
+ * from the hook in the previous entry.
+ */
+class cache_hook {
+public:
+    cache_hook() noexcept = default;
+
+    /**
+     * Copy constructor.
+     *
+     * The value of evicted() will be false for the newly constructed hook.
+     */
+    cache_hook(const cache_hook&) = default;
+
+    cache_hook& operator=(const cache_hook&) = delete;
+    cache_hook(cache_hook&&) noexcept = delete;
+    cache_hook& operator=(cache_hook&&) noexcept = delete;
+    ~cache_hook() noexcept = default;
+
+    /**
+     * Mark a notable interaction with the entry containing this hook.
+     */
+    void touch() noexcept { freq_ = std::min(freq_ + 1, 3); }
+
+    /**
+     * Check if an entry containing this hook is in the cache.
+     *
+     * This method only considers the small and main cache queues, and does not
+     * reflect membership on the ghost queue. This method will return true for a
+     * default constructured hook.
+     *
+     * An entry need not check if an entry has been evicted using this
+     * interface, and may instead arrange for a side-effect of the cache's
+     * configured eviction function to mark an entry as evicted. Consider for
+     * example the following entry and eviction function:
+     *
+     *     struct entry { std::unique_ptr<int> ptr; }
+     *
+     *     struct evict {
+     *         bool operator()(entry& e) {
+     *             e.ptr = nullptr;
+     *             return true;
+     *         }
+     *     };
+     *
+     * An application may then use `e.ptr` to indicate that a entry has been
+     * evicted instead of invoking cache_hook::evicted.
+     */
+    [[nodiscard]] bool evicted() const noexcept { return !hook_.is_linked(); }
+
+private:
+    using hook_type = boost::intrusive::list_member_hook<
+      boost::intrusive::link_mode<boost::intrusive::safe_link>>;
+
+    /*
+     * the location is checked only in cache::erase where the case of the hook
+     * being on neither main nor small is indicated by evicted() being true.
+     */
+    enum class location : uint8_t { main, small };
+
+    template<
+      typename T,
+      cache_hook T::*Hook,
+      cache_evictor<T> Evictor,
+      cache_cost<T> Cost>
+    friend class cache;
+
+    friend class testing_details::cache_hook_accessor;
+
+    uint8_t freq_{0};
+    std::optional<uint64_t> insertion_time_;
```

Comment:
Do you think it is worth using `0` to indicate absence of insertion_time_ and avoid the optional?

Reply:
Maybe, but using the optional means I don't have to think / worry about the initialization state of the clock and its affect on how a default-constructed hook is evaluated in terms of is membership on the ghost queue.

More broadly speaking, the insertion time handling can (1) be merged with `freq_` and we can use a bit in there for the optional state, and the ultimate form this will take is to make the insertion time compile-time optional--it isn't really needed if the indexing structure that an application uses can maintain a fifo ordering without using time.

Reply:
> it isn't really needed if the indexing structure that an application uses can maintain a fifo ordering without using time

I sort of wonder if we should have a cache class for each of the "common" indexing structures (btree_map, chunked_fifo, etc) so that isn't repeated all over the place, but again, probably worth it to see this class in action a few times before centralizing anything.

Reply:
@rockwotj 

yes, great idea--actually was just thinking about this on my run as a convenience utility. but let me give a concrete example of where this breaks down:

In the page cache we have a global eviction policy--i want to find the page out of all pages to evict. But the indexing strategy is that pages are bound to files.

So if we have a single index, then we need keys that look like "file.page-number" or something like this. However, structurally we already represent the necessary disaggregation. For example, a segment is represented in memory and can hold its own `btree_map` for the pages of its file, allowing for simple integer page number index entries. In this case we do not have a single index structure, but we do have a single eviction strategy across all pages.

Reply:
Yeah totally, I do like the flexibility of this design and it totally makes sense in the context you explained. We'll have to watch the other places this is used and maybe there will be a common pattern for "map like cache" that we'll want a class for it that uses this eviction algorithm. All and all, this LGTM and I'm sure we'll feel the cost of adding this structure multiple times and be ready to add in this if there is a need